### DISCUSSION TITLE
Vitest auto-imports from Nuxt3 and Vueuse

### DISCUSSION THREAD

**Discussion by jimbo0378** [vitest-dev/vitest#1737] [Q&A]
------------------------------
- Operating System: `Linux`
- Node Version:     `v16.15.1`
- Nuxt Version:     `3.0.0-rc.4-27600153.ad42023`
- Package Manager:  `npm@8.11.0`
- Builder:          `vite`
- User Config:      `app`, `meta`, `css`, `ssr`, `vite`, `vue`, `runtimeConfig`, `modules`
- Runtime Modules:  `@vueuse/nuxt@8.7.5`
- Build Modules:    `-`
------------------------------

Hi all, we're trying to get Vitest running in our Nuxt3 application. 
But we're not able to get the auto-imports from Nuxt (and vueuse) working without specifying them inside the `unplugin-auto-import/vite` config and mocking the methods, but this isn't maintainable. Is there a way to get these auto-imports to work without this workaround?
I've seen [this post](https://github.com/vitest-dev/vitest/discussions/1565), but it doesn't fix the Nuxt auto-imports issue, only for Vue imports.

**vitest.config.js**

```
import { defineConfig } from 'vitest/config';
import Vue from '@vitejs/plugin-vue';
import AutoImport from 'unplugin-auto-import/vite';
import Components from 'unplugin-vue-components/vite';
import { alias } from './tests/alias';

export default defineConfig({
    plugins: [
        Vue({
            template: {
                compilerOptions: {
                    isCustomElement: tag => tag.includes('-'),
                },
            },
        }),
        Components({
            dirs: ['./components'],
        }),
        AutoImport({
            imports: [
                'vitest',
                'vue',
                {
                    '#imports': [
                        'useNuxtApp',
                        'useBaseFetch',
                        'useRuntimeConfig',
                        'useState',
                        'useLazyAsyncData',
                        'useLocalisationOptions',
                        'useFavoriteLocalisation',
                        'useInputValidation',
                    ],
                },
            ],
        }),
    ],
    test: {
        environment: 'jsdom',
        include: ['**/*.{test,spec}.{js,ts}'],
        deps: {
            inline: [/@nuxt\/test-utils-edge/],
        },
        coverage: {
            reporter: ['text', 'json', 'html'],
        },
        globals: true,
        setupFiles: './tests/setup.ts',
    },
    resolve: {
        alias,
    },
});
```

**setup.ts**

```
vi.mock('#imports', () => {
    return {
        useNuxtApp: vi.fn().mockImplementation(() => ({
            $user: ref(new User({ data: userWithPersonData, pending: false, error: false })),
        })),
        // ....
    };
});
```

**Comment by sebalaini** [comment]
having the same issue with auto import functions like `useCookie`

**Reply by dudintv** [reply]
and. the same thing with `useRoute`, `useRouter`

**Comment by floriankapaun** [comment]
Has anyone found a solution yet? I'm having the problem using the current latest versions of nuxt (`3.2.2`) and everything else. 🤔

**Reply by dsvgl** [reply]
Check out https://github.com/danielroe/nuxt-vitest

**Reply by sebalaini** [reply]
The above is a wrapper of vitest to work with Nuxt? BTW seems to be in beta and it say:

Warning This library is in active development and you should pin the patch version before using.

**Comment by sebalaini** [comment]
I managed to get Vue specific stuff but not Nuxt.

```
  plugins: [
    vue(),
    AutoImport({
      imports: ['vue'],
      dirs: ['./composables/']
    })
```

**Comment by adamskeeled** [comment]
Anyone managed to get this working for Nuxt auto-imports?

**Reply by OleksandrAfonichev** [reply]
it works fine with https://github.com/danielroe/nuxt-vitest

but because of it there are other problems that I can not solve yet. For example:
<img width="918" alt="Screenshot 2023-03-23 at 17 08 20" src="https://user-images.githubusercontent.com/118198655/227265044-2b00884f-ff2e-420b-b78e-1c69c223c4b8.png">
(this happens when i use `<transition>` in my code)
if you manage to solve this problem, let me know

**Comment by uke5tar** [comment]
Yes. I got it working with Nuxt3 (auto import) but haven't tried it with vueuse yet. Here is the trick for nuxt3 auto import:

install and import those dependencies in your vitest.config.ts:
```
import vue from '@vitejs/plugin-vue'
import AutoImport from 'unplugin-auto-import/vite'
import Components from 'unplugin-vue-components/vite'

export default defineConfig({
  plugins: [
    vue(),
    Components({
      dirs: ['./components'] // this will auto import everything from within nuxt
    }),
    AutoImport({
      imports: [
        'vue',
        'vitest',
        'pinia' // if u use pinia
      ],
      dirs: ['./composables', './stores', './utils'], // choose whatever folders u want to auto import
      dts: true
    })
  ],
})

If u check the unplugin Autoimport docs there is also a part about vueuse. Should be pretty straight forward: https://github.com/antfu/unplugin-auto-import

**Reply by v3nt** [reply]
@uke5tar I still get issues with `useHead`, `useSeoMeta`, `useRoute` etc. Did you get these working?
```ReferenceError: useRoute is not defined```

**Reply by uke5tar** [reply]
Those composables are autoimported for me in Nuxt3 without needing to manually configure something. Are you using Nuxt3?

**Reply by v3nt** [reply]
yes were using Nuxt 3. Tried 3 setups from scratch and same issue.

**Reply by v3nt** [reply]
ie here is the issue reproduced. 
https://github.com/divasilevski/nuxt-starter/issues/9

**Reply by uke5tar** [reply]
This seems weird. This should work out of the box in Nuxt3. What if you create a new example project somewhere with Nuxt3 and add useHead or useRouter? It should work out of the box.

Or is this only a ts error for you? Then you can add in your tsconfig.json
```
{
  "extends": "./.nuxt/tsconfig.json",
  }
}
```

**Reply by v3nt** [reply]
thanks but tsconfig already has that.

**Reply by uke5tar** [reply]
Then all I can say to you is that useHead and useRoute should be 100% working as autoimport out of the box with nuxt3 unless you disabled autoimport https://nuxt.com/docs/guide/concepts/auto-imports#disabling-auto-imports.

You can create a fresh project, confirm that autoimport is working and then maybe step by step add your config until you reach the error.

**Comment by v3nt** [comment]
Does anyone know how to auto import ./plugins/** ? Tried the above but get the same error.

### Plugin function in component 
This is accessibly globally without the need for imports in components.
`<div v-if="$isAdminModeUI()"></div>`

### Plugins
<img width="866" alt="image" src="https://github.com/vitest-dev/vitest/assets/6348562/7ad1b9de-37f7-4a2d-acc6-bd19724542c5">

### Vitest config
<img width="530" alt="image" src="https://github.com/vitest-dev/vitest/assets/6348562/7531a01e-c38f-4336-8c2c-afcbf77c93e8">

**Comment by patrickyue** [comment]
Have you tried putting those into nuxt.config.ts?
My issue is that I can't use any global composables.
But we solved this by creating a nuxt.config.ts on the root folder, which vitest.config.ts is there too
and inside that nuxt.config, import those global dirs / add global modules as nuxt-vitest , @vueuse/nuxt ,etc

![image](https://github.com/vitest-dev/vitest/assets/8323205/43ab766d-4838-44fc-9b87-874a5acedaf0)

**Reply by simon-kramer** [reply]
Got the same issue as all of these ppl, will try this approach & come back to report :)
